WIP: Added support for more tado devices

wmalgadey opened the pull request:
## Description:
Added some functionality to support AC devices and HOT WATER devices

**Related issue (if applicable):** fixes #6943

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

Commit:
added heating and cooling functionality

* requierements to pytado 0.2.5

Commit:
fixed errors and changed capabilities cases

Yippy committed:
convert_temperature function is not required here

Yippy committed:
convert_temperature function is required here, must import class

Yippy committed:
Added constant of zones
Added sensors for air conditioning, such as temperature, humidity and power on

Yippy committed:
Rename ac_mode to device_type, as Tado will populate it as "HEATING" or "AIR_CONDITIONING".
Added ac_mode to set the correct mode for api call, this parameter is either "HEAT" or COOL"

Yippy committed:
Support toggling 'ON' and 'OFF' for HOT_WATER device type, temperature would normally be 'None' check device_is_active and switch power.
Use wmalgadey/PyTado 2.5 release

Yippy committed:
Fix issue with API with 'HEAT' and 'COOL'
Made 'min_temp' and 'max_temp' variable as dictionary where modes allow for different modes 'HEAT' or 'COOL'
Remove '_is_cooling' variable as this was not needed, the operation mode will set the mode of 'HEAT' or 'COOL'

Yippy committed:
Ensure multiple temperature of 'HEAT' or 'COOL' to be stored
Add HOT_WATER to controllable climate.

Yippy committed:
Display HOT_WATER manual power on operation list

Yippy committed:
Change requirement to python-tado 0.2.6

Commit:
Merge pull request #1 from Yippy/tado_ac_sensor

Add Tado AIR_CONDITIONING sensor

homeassistant commented:
Hi @Yippy, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#18331).

Once you do that we will be able to review and accept this pull request.

Thanks!

Commit:
fixed intendation and line length

homeassistant commented:
Hi @Yippy, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#18331).

Once you do that we will be able to review and accept this pull request.

Thanks!

Commit:
fixed more line length errors

homeassistant commented:
Hi @Yippy, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#18331).

Once you do that we will be able to review and accept this pull request.

Thanks!

Commit:
fixed tox errors

Commit:
wrong return value removed

Commit:
reverted b792a43 and added else-statement

Commit:
pylint error fixed

wmalgadey commented:
can somebody official tell us, why the python-tado lib (PyTado) from pypi doesn't work in the pylint-tests on travis?

`E0401: Unable to import 'PyTado.interface' (import-error)`

Commit:
Fixed an error in python-tado package

wmalgadey commented:
> can somebody official tell us, why the python-tado lib (PyTado) from pypi doesn't work in the pylint-tests on travis?
> 
> `E0401: Unable to import 'PyTado.interface' (import-error)`

Somebody gave me a big hint on why the package doesn't work. I changed the setup script of python-tado, so not everything was included in the package as before! I changed to 0.2.7 where I included everything right in the package.

kloknibor committed:
Fixing hot water

With the old solution hot water was only able to switch to a temp or an active mode when it came from the off state. (Since as soon as device became active power would be OFF so you were never able to switch temp only when device was not active)

Commit:
Merge pull request #2 from kloknibors-forking-place/tado_ac

Fixing hot water

* I can't test this, because I have no hot water device from tado

homeassistant commented:
Hi @kloknibor, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#18331).

Once you do that we will be able to review and accept this pull request.

Thanks!

Commit:
fixed pylint errors

homeassistant commented:
Hi @kloknibor, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#18331).

Once you do that we will be able to review and accept this pull request.

Thanks!

Commit:
fixed trailing whitespace

homeassistant commented:
Hi @kloknibor, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#18331).

Once you do that we will be able to review and accept this pull request.

Thanks!

Commit:
fixed blank lines

balloob commented on homeassistant/components/climate/tado.py:
```diff
@@ -160,26 +203,40 @@ def current_temperature(self):
     def operation_list(self):
         """Return the list of available operation modes (readable)."""
-        return list(OPERATION_LIST.values())
+        operations = list(OPERATION_LIST.values())
```
This is all wrong. Operation modes should be limited to the states as provided by the Home Assistant Climate component: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/climate/__init__.py#L43-L50

No other modes are allowed because it breaks the integration with Alexa, Google Assistant etc.

balloob replied:
Okay, so let's close this PR then?

wmalgadey replied:
??? so you want to remove the tado integration from hass!?

Yippy replied:
@balloob
STATE_HEAT = 'heat'
STATE_COOL = 'cool'
STATE_IDLE = 'idle'
STATE_AUTO = 'auto'
STATE_MANUAL = 'manual'
STATE_DRY = 'dry'
STATE_FAN_ONLY = 'fan_only'
STATE_ECO = 'eco'

Tado do not use those singular states, if you want to revamp Hass to cater tado it need both heat and cool states

Heat_auto
Heat_manual
Heat_timer
Heat_off

Cool_auto
Cool_manual
Cool_timer
Cool_off

Due to the fact tado can handle both hot and cold not just hot in many thermostat

balloob replied:
Nobody will turn on the heating and the cooling at the same time so that can easily be represented in a single operation mode.

I am not discussing this, I don't have time for that. PRs that do not follow our states are not accepted, period. 

So shall I close this PR?

wmalgadey replied:
I wouldn't discuss this, if it would be a new component. But this is only a change to an existing one!

We change the PR to WIP and look if we can integrate the default states. I would guess that `timer` and `manual` are nearly the same in tado and `auto` is the tado schedule mode. So we may need a translation for hass states to tado states.

kloknibor replied:
I suppose we could reuse it like :

STATE_AUTO = 'auto' = tado_mode
STATE_MANUAL = 'manual' = manual
STATE_ECO = 'eco' = smart schedule

Would make it more confusing for the users but we keep all of the current functionality and I suppose we would Follow homeassistant states

wmalgadey replied:
we can leave it nearly the same for the user, I suppose! We already did a "translation" for the UI and the Tado States. So now, we may limit the states to those in the climate devices and map the UI to those states. After that we map the climate devices steps to tado states (this should best be done inside the method, where we actually communicate with tado/PyTado)

wmalgadey replied:
And "timer" is like
-> "manual" + switch back to "tado schedule" after a timespan.
  or
-> "manual" + switch back to "tado schedule" when tado would change it next.

whereas "manual" is like
-> "manual" + never switch back

and "auto"
-> "tado schedule mode"

So we may have to do more checks, what is the current tado state and set tado states accordingly. Currently we are doing this a bit too simple, because we just switch to "manual" and set the temperature, we do not change the tado "manual" properties like the time when it will switch back or if it should wait until tado switches it back manually.

kloknibor replied:
Currently manual just means that the user needs manually stop the overlay. Since manual is an overlay right? And there is no option to add a time for the timer from the UI which basically renders it useless unless we hardcode this time?

Sorry I think I kind of miss your point/Don't understand?

kloknibor replied:
I guess my point was that we don’t really need to change the behavior 😜 We just need to rename the tado overlays to the modes which are used within homeassistant right?

balloob commented on homeassistant/components/climate/tado.py:
```diff
@@ -160,26 +203,40 @@ def current_temperature(self):
+            operations.extend(list(OPERATION_MANUAL.values()))
+        elif self._device_type == "HOT_WATER" and \
+                not self._is_manual_temperature:
+            operations.extend(list(OPERATION_MANUAL_HOT_WATER_ON.values()))
```
This should not be in a climate component to begin with, there is a water heater component for this.

kloknibor replied:
At the current state off the hot_water component probably not... I would need to change the hot water device to get this working, like it is working right now.

Which would bring it closer to the climate device again which I’m not sure if desired. Also we will have the issue with the non default home-assistant modes.

@balloob what is the reason to split the hot_water? Because it seems to work tightly with your Heating since it’s the same system?

balloob commented:
Closing old WIP branches.
